### PR TITLE
feat(tracking): add tracking on ods components

### PR DESCRIPTION
**Pull request by aboungnaseng-ovhcloud**
ref: MANAGER-11442

| Question         | Answer
| ---------------- | ---
| Branch?          | `master` / `release/**` / `develop` 
| Bug fix?         | yes/no
| New feature?     | yes/no
| Breaking change? | yes/no
| Tickets          | Fix #... 
| License          | BSD 3-Clause

- [ ] Try to keep pull requests small so they can be easily reviewed.
- [ ] Commits are signed-off
- [ ] Only FR translations have been updated
- [ ] Branch is up-to-date with target branch
- [ ] Lint has passed locally
- [ ] Standalone app was ran and tested locally
- [ ] Ticket reference is mentioned in linked commits (internal only)
- [ ] Breaking change is mentioned in relevant commits

### CODE REVIEW

File: packages/manager/core/application/src/ovh-tracking.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        page: { name: page },
+        complete_page_name: page,
+        level2: '81',
+        type: 'action',
+      });
+    });
+  };
+
+  // Send tracking click for the ODS_COMPONENTS_GROUP
+  const ovhTrackingParenNode = (event) => {
```
Comment (anooparveti) [inline packages/manager/core/application/src/ovh-tracking.tsx:66]:
```suggestion
  const ovhTrackingParentNode = (event) => {
```

File: packages/manager/core/application/src/ovh-tracking.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      if (tabsEvents.length > 0)
+        tabsEvents.forEach((element) => {
+          element.removeEventListener(
+            'odsValueChange',
+            ovhTrackingParenNode,
+            false,
+          );
+        });
+    };
+  }, [location]);
```
Comment (anooparveti) [inline packages/manager/core/application/src/ovh-tracking.tsx:108]: **question:**  Default tracking is added to ODS components when the location changes. But what if some ODS components are added to the DOM after some user actions on the current screen.

For ex,
1. User lands on the config page
2. Enters the value in `Field A`
3. Based on the response to `Field A`, we decide to display either `Field B` or `Field C`.

So with this implementation, are the fields displayed conditionally are not tracked right as there is no change in the location?

File: packages/components/ovh-at-internet/src/ovh-at-internet.ts [hunk 3]
```diff
@@ -297,13 +307,25 @@ export default class OvhAtInternet extends OvhAtInternetConfig {
...
+        const pageTrackingData2 = getPageTrackingData(data?.page);
+        tracking = {
+          ...tracking,
+          page: pageTrackingData2.page,
+          page_chapter1: pageTrackingData2.page_chapter1,
+          page_chapter2: pageTrackingData2.page_chapter2,
+          page_chapter3: pageTrackingData2.page_chapter3,
+          page_complete: data?.page?.name,
+        };
+      }
```
Comment (anooparveti) [inline packages/components/ovh-at-internet/src/ovh-at-internet.ts:327]: **question:** Can you please explain me this? Why do we need to execute `getPageTrackingData` on `data` as well as `data.page`?

Comment (aboungnaseng-ovhcloud) [reply packages/components/ovh-at-internet/src/ovh-at-internet.ts:327]: sure i will update it

File: packages/components/ovh-at-internet/src/ovh-at-internet.ts [hunk 4]
```diff
@@ -115,6 +118,10 @@ export default class OvhAtInternet extends OvhAtInternetConfig {
   }
 
   shouldUsePianoAnalytics() {
+    if (window.location?.hostname === 'localhost') {
+      return true;
+    }
```
Comment (anooparveti) [inline packages/components/ovh-at-internet/src/ovh-at-internet.ts:123]: Use environment variable also to enable tracking in localhost.
Environment variable can be defined in `packages/manager/tools/dev-server/bin/manager-dev-server.js`

**Review by chipp972 — APPROVED** [review]

File: packages/manager/core/application/src/ovh-routing.tsx [hunk 5]
```diff
@@ -36,6 +45,16 @@ function HidePreloader(): JSX.Element {
 
 function OvhContainerRoutingSync(): JSX.Element {
   const location = useLocation();
+  const outlet = useOutlet();
+  const navigate = useNavigate();
+  useEffect(() => {
+    setTimeout(() => {
+      const child = outlet.props?.children?.props?.children?.props?.children;
```
Comment (anooparveti) [inline packages/manager/core/application/src/ovh-routing.tsx:52]: **question:** Why is this nesting required?
Also, this seems like a bad approach as `props.childern` can have multiple levels of children

File: packages/manager/core/application/src/ovh-tracking.tsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    const path = myStateRef.current.pathname.split('/')[1];
+    env.then((response) => {
+      const { applicationName, universe } = response;
+      const name = `${applicationName}::${path || 'homepage'}::${value}`;
+      const page = `${universe}::${applicationName}::${path || 'homepage'}`;
+      tracking.trackClick({
+        name,
+        page: { name: page },
+        complete_page_name: page,
+        level2: '81',
```
Comment (anooparveti) [inline packages/manager/core/application/src/ovh-tracking.tsx:59]: **nitpick:** `level2` is set based on the universe. So it should not be hard-coded

**Review by anooparveti — CHANGES_REQUESTED** [review]

File: packages/manager/core/application/src/ovh-tracking.tsx (hunk 6 above)
Comment (aboungnaseng-ovhcloud) [reply packages/manager/core/application/src/ovh-tracking.tsx:59]: For the moment, we didn't have the key in the useShell(), that why I hard coded this

In the this file 'packages/components/ovh-at-internet/src/constants.ts' you have a constant AT_INTERNET_LEVEL2, but we can't get the key nowhere

File: packages/manager/core/application/src/ovh-routing.tsx [hunk 7]
```diff
@@ -107,16 +115,29 @@ export function createAppRouter() {
...
-  const routes = [...regularRoutes, { path: '*', element: <NotFound /> }];
+  const newRoutes = regularRoutes.map((route, index) => {
+    if (route.path === 'dashboard') {
+      return {
+        ...route,
+        path: '*',
+        errorElement: <NotFound />,
+        error: <NotFound />,
+      };
+    }
```
Comment (anooparveti) [inline packages/manager/core/application/src/ovh-routing.tsx:126]: **question:** Why is it different for `dashboard`?

File: packages/manager/core/application/src/ovh-tracking.tsx [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+  const OvhTrackPage = (loc: any) => {
+    const path = loc?.pathname.split('/')[1];
+    const partsReplace = loc.pathname.substring(1).replaceAll('/', '::');
+    env.then((response) => {
+      const { applicationName, universe } = response;
+      const name = `${universe}::${applicationName}::${partsReplace ||
+        'homepage'}`;
+      tracking.trackPage({
+        name,
+        level2: '81',
```
[the remark this answers is not in the document]
Comment (anooparveti) [reply packages/manager/core/application/src/ovh-tracking.tsx:32]: If the key is not present in `useShell()`, you can add it.

File: packages/manager/core/application/src/ovh-tracking.tsx (hunk 6 above)
Comment (anooparveti) [reply packages/manager/core/application/src/ovh-tracking.tsx:59]: If the key is not present in `useShell()`, you can add it.

File: packages/manager/core/application/src/ovh-routing.tsx (hunk 7 above)
Comment (aboungnaseng-ovhcloud) [reply packages/manager/core/application/src/ovh-routing.tsx:126]: Its different for dashboard to stay iso as angular, to have the services name in first path 

https://www.ovh.com/manager/#/dedicated/nasha/{service}

and not  :

https://www.ovh.com/manager/#/dedicated/nasha/dashboard/{service}

the package generouted didn't allow us to us parameter in first path so we need to made a fix

**Pull request merged**